test/charms/test_main/src/charm.py

```diff
@@ -138,6 +141,12 @@ def on_foo_bar_action(self, event):
         self._state['observed_event_types'].append(type(event))
         self._write_state()
 
+    def on_collect_metrics(self, event):
+        self._state['on_collect_metrics'].append(type(event))
+        self._state['observed_event_types'].append(type(event))
+        event.add_metrics({'foo': 42}, {'bar': 4.2})
```

Comment:
Is it required for add-metrics to be called inside the collect-metrics hook? What happens if it's called by other hooks?

Reply:
```ERROR cannot record metric: metrics not allowed in this context```

Reply:
So, yes, it is required currently to do so in a collect-metrics hook much like with actions because there is a separate context.